lib/internal/streams/operators.js

```diff
@@ -0,0 +1,269 @@
+'use strict';
+
+const { AbortController } = require('internal/abort_controller');
+const { isIterable } = require('internal/streams/utils');
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_ARG_VALUE,
+  }, AbortError,
+} = require('internal/errors');
+
+const {
+  NumberIsInteger,
+  Promise,
+  PromiseReject,
+  Symbol,
+} = primordials;
+
+const kEmpty = Symbol('kEmpty');
+
+async function * map(fn, options) {
+  if (typeof fn !== 'function') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'fn', ['Function', 'AsyncFunction'], this);
+  }
+
+  let concurrency = 1;
+  if (options?.concurrency != null) {
+    concurrency = options.concurrency;
+  } else if (options != null && typeof options !== 'object') {
+    concurrency = options;
+  }
+
+  if (!NumberIsInteger(concurrency)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'concurrency', ['Integer'], concurrency);
+  }
+
+  if (concurrency < 1) {
+    throw new ERR_INVALID_ARG_VALUE('concurrency', concurrency);
+  }
+
+  const ac = new AbortController();
+  const stream = this;
+  const queue = [];
+  const signal = ac.signal;
+  const signalOpt = { signal };
+
+  options?.signal?.addEventListener('abort', () => ac.abort(), { weak: true });
+
+  let next;
+  let resume;
+  let done = false;
+
+  function onDone() {
+    done = true;
+  }
+
+  async function pump() {
+    try {
+      for await (let val of stream) {
+        if (done) {
+          return;
+        }
+
+        if (signal.aborted) {
+          throw new AbortError();
+        }
+
+        try {
+          val = fn(val, signalOpt);
+        } catch (err) {
+          val = PromiseReject(err);
+        }
+
+        if (val === kEmpty) {
+          continue;
+        }
+
+        if (typeof val?.catch === 'function') {
+          val.catch(onDone);
+        }
+
+        queue.push(val);
+        if (next) {
+          next();
+          next = null;
+        }
+
+        if (!done && queue.length >= concurrency) {
+          await new Promise((resolve) => {
+            resume = resolve;
+          });
+        }
+      }
+      queue.push(null);
+    } catch (err) {
+      const val = PromiseReject(err);
+      val.catch(onDone);
+      queue.push(val);
+    } finally {
+      done = true;
+      if (next) {
+        next();
+        next = null;
+      }
+    }
+  }
+
+  pump();
+
+  try {
+    while (true) {
+      while (queue.length > 0) {
+        let val = queue[0];
+
+        if (val === null) {
```

Comment:
From what I understand, you can't map a stream value to null then? Unless you `async () => null` or something?

Reply:
@gioragutt these sort of comments would be better as test cases :)

Reply:
`Readable` doesn't support null as a value unfortunately.

Reply:
I've changed the operators to support null but it will break in `Readable.from`.